### ISSUE TITLE
7.6.1 breaks gradle-consistent-versions

### ISSUE THREAD

**Issue by CRogers** [gradle/gradle#24234]
### Expected Behavior
[`gradle-consistent-versions`](https://github.com/palantir/gradle-consistent-versions#compalantirconsistent-versions---) is one of our most essential gradle plugins used by 1000s of repos internally at our company.

One half of the plugin takes a `versions.props` file ([example](https://github.com/palantir/conjure-java-runtime/blob/51f4e0a39c945d9423c50d512fffab21fb2665b1/versions.props)) and applies these constraints to all configurations in all projects. The expectation is that if you add a constraint to this `versions.props`, it will end up in all your configurations (that it makes sense to apply constraints automatically to).

I have a made [a minimal reproducer repository](https://github.com/CRogers/gradle-7.6.1-gcv-repro) to demonstrate this. It has [a project with two configurations that are resolved sequentially by a task](https://github.com/CRogers/gradle-7.6.1-gcv-repro/blob/ded8b07f1e87f21559533d00cf0d7cf5f15e27fc/build.gradle#L37-L52), pulling in the [constraints from `versions.props`](https://github.com/CRogers/gradle-7.6.1-gcv-repro/blob/ded8b07f1e87f21559533d00cf0d7cf5f15e27fc/versions.props) using gradle-consistent-versions:

```
# versions.props
com.google.guava:guava = 31.1-jre
junit:junit = 4.12
```

```gradle
// build.gradle snippet
configurations {
    one
    two
}

dependencies {
    one 'com.google.guava:guava'
    two 'junit:junit'
}

task resolveBoth {
    doFirst {
        configurations.one.resolve()
        configurations.two.resolve()
    }
}
```

With Gradle 7.6, if you run `./gradlew resolveBoth` both configurations manage to resolve as they both have the constraints loaded from `versions.props`. See this [buildscan when running with Gradle 7.6](https://scans.gradle.com/s/ndp5y5z2yhr3g/dependencies?toggled=W1swXSxbMCwwXSxbMCwxXV0):

![Screenshot 2023-03-08 at 16 40 32](https://user-images.githubusercontent.com/397795/223774410-60ca6643-591c-4c35-8dd0-04cfa5331f58.png)

### Current Behavior
With Gradle 7.6.1, this same code does not manage to resolve the second configuration (or any configuration after the first). See this [buildscan when running with Gradle 7.6.1](https://scans.gradle.com/s/x52zvtf2qdd4o/dependencies?toggled=W1swXSxbMCwxXSxbMCwwXV0):

![Screenshot 2023-03-08 at 16 43 32](https://user-images.githubusercontent.com/397795/223775120-d26a1df9-eeee-47fa-917d-e274806d50b1.png)

Unfortunately, this causes every repo using gradle-consistent-versions (which is 99% of our repos) to not work with Gradle 7.6.1 (and by extension 8.0.2).

#### Regressed commit in Gradle

I ran a git bisect between `v7.6.0` and `v7.6.1` and found this to be the commit that causes the regression:

* ["Re-evaluate component configuration metadata if the underlying configuration may have changed"](https://github.com/gradle/gradle/commit/46e39a311d3909428419ca835545ce7a63480487) by @big-guy / @ljacomet (from ["Reduce memory consumption more" PR](https://github.com/gradle/gradle/pull/23609)).

Buildscans before and at this commit:
* Before @ https://github.com/gradle/gradle/commit/f492a51d9984915f5e0582c940be4a7b669572a5: https://scans.gradle.com/s/4kj22kbazgfss
* At regression @ https://github.com/gradle/gradle/commit/46e39a311d3909428419ca835545ce7a63480487: https://scans.gradle.com/s/4cpltyjyzkif2

#### gradle-consistent-versions Internals Context
gradle-consistent-versions achieves this constraint propagation by creating a ["platform" configuration of dependency constraints in the root project](https://github.com/palantir/gradle-consistent-versions/blob/f407769c01dba863147e6108bd12e6990ab2ebed/src/main/java/com/palantir/gradle/versions/VersionsPropsPlugin.java#L95-L104). Another root project configuration [`rootConfiguration`](https://github.com/palantir/gradle-consistent-versions/blob/f407769c01dba863147e6108bd12e6990ab2ebed/src/main/java/com/palantir/gradle/versions/VersionsPropsPlugin.java#L116-L119) then [adds the constraints from the "platform" configuration using a project dependency](https://github.com/palantir/gradle-consistent-versions/blob/f407769c01dba863147e6108bd12e6990ab2ebed/src/main/java/com/palantir/gradle/versions/VersionsPropsPlugin.java#L135-L142). (Pretty much) all other configurations [extendsFrom this `rootConfiguration`](https://github.com/palantir/gradle-consistent-versions/blob/f407769c01dba863147e6108bd12e6990ab2ebed/src/main/java/com/palantir/gradle/versions/VersionsPropsPlugin.java#L213) to ensure they get all the constraints.

You'll notice [the regressed 7.6.1 scan](https://scans.gradle.com/s/x52zvtf2qdd4o/dependencies?focusedConfiguration=WzAsMSxbMCwxXV0&toggled=W1swXSxbMCwxXSxbMCwwXV0) is missing the [reference to the root project constraint "platform" the successful build using 7.6 had](https://scans.gradle.com/s/ndp5y5z2yhr3g/dependencies?focusedDependency=WzAsMSwxLFswLDEsWzFdXV0&focusedDependencyView=details&toggled=W1swXSxbMCwwXSxbMCwxXV0):

![Screenshot 2023-03-08 at 16 46 42](https://user-images.githubusercontent.com/397795/223775971-1977b53b-6cf6-4a95-9832-7d8c0de3d743.png)

### Steps to Reproduce 
Run `./gradle resolveBoth` on https://github.com/CRogers/gradle-7.6.1-gcv-repro/commit/b66a044e398538844c5e5c3176cae84b1ddb9136. The [previous commit](https://github.com/CRogers/gradle-7.6.1-gcv-repro/commit/ded8b07f1e87f21559533d00cf0d7cf5f15e27fc) uses 7.6 and works.

Additionally gradle-consistent-versions has [a branch](https://github.com/palantir/gradle-consistent-versions/commit/9e4a55e35b3c88f067e38dff421e0e9f6e1cee93) showing it's [tests succeeding with 7.6 and failing with 7.6.1](https://app.circleci.com/jobs/github/palantir/gradle-consistent-versions/12948).

### Your Environment
Build scan URLs:
* 7.6: https://scans.gradle.com/s/ndp5y5z2yhr3g
* 7.6.1: https://scans.gradle.com/s/x52zvtf2qdd4o

**Comment by big-guy** [comment]
@CRogers thanks for the detailed report. We'll take a look

**Comment by jvandort** [comment]
@CRogers It looks like this may have been fixed in 8.0.2 and as a result 8.1 (nightly) already by [this PR](https://github.com/gradle/gradle/pull/24118). Would you be able to test with either of these versions to confirm or deny whether this is fixed there? 

If so, we can be sure to target the same fix to a 7.x backport.

**Comment by CRogers** [comment]
@jvandort I can confirm it:

* [Doesn't work in 8.0.1](https://scans.gradle.com/s/g4dgxx34wgvlm)
* [Works in 8.0.2](https://scans.gradle.com/s/6kcotxwgglfvm)

**Comment by CRogers** [comment]
FWIW I also tested against that commit and the parent commit and can confirm it's specifically that which fixes it:

* Fix commit 79a3a1471a4 succeeds: https://scans.gradle.com/s/kume5ghsxogoq
* Parent of above 038ae5945c2 fails: https://scans.gradle.com/s/l7wamyxawwkyi

**Comment by jvandort** [comment]
@CRogers Thank you for verifying this. 

Lets leave this issue open and targeted for `7.x backport`, so that if/when we release a 7.7, we can be sure https://github.com/gradle/gradle/commit/79a3a1471a426a67e2e53d9076b8c52c2eef9abf is included.

**Comment by ljacomet** [comment]
Resolved with #25166
